crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -783,64 +790,124 @@ where
-        for archived_segment in older_archived_segments {
-            send_archived_segment_notification(
-                &archived_segment_notification_sender,
-                archived_segment,
-            )
-            .await;
-        }
+        let mut block_importing_notification_stream = block_importing_notification_stream;
 
-        while let Some(ref block_import_notification) =
-            block_importing_notification_stream.next().await
-        {
-            archive_block(
-                &mut archiver,
-                segment_headers_store.clone(),
-                client.clone(),
-                &sync_oracle,
-                telemetry.clone(),
-                archived_segment_notification_sender.clone(),
+        'initialization_recovery: loop {
+            let overridden_last_archived_block = saved_block_import_notification
+                .as_ref()
+                .map(|notification: &BlockImportingNotification<Block>| notification.block_number);
+            let archived_segment_notification_sender =
+                subspace_link.archived_segment_notification_sender.clone();
+
+            let archiver = match maybe_archiver.take() {
+                Some(archiver) => archiver,
+                None => initialize_archiver(
+                    &segment_headers_store,
+                    &subspace_link,
+                    client.as_ref(),
+                    overridden_last_archived_block,
+                )?,
+            };
+
+            let InitializedArchiver {
                 confirmation_depth_k,
-                &mut best_archived_block_number,
-                &mut best_archived_block_hash,
-                block_import_notification.block_number,
-            )
-            .await?;
-        }
+                mut archiver,
+                older_archived_segments,
+                best_archived_block: (mut best_archived_block_hash, mut best_archived_block_number),
+            } = archiver;
+
+            debug!(%best_archived_block_number, ?best_archived_block_hash,  "Archiver initialized.");
+
+            if let Some(block_import_notification) = saved_block_import_notification.take() {
+                let success = archive_block(
+                    &mut archiver,
+                    segment_headers_store.clone(),
+                    client.clone(),
+                    &sync_oracle,
+                    telemetry.clone(),
+                    archived_segment_notification_sender.clone(),
+                    confirmation_depth_k,
+                    &mut best_archived_block_number,
+                    &mut best_archived_block_hash,
+                    block_import_notification.block_number,
+                )
+                .await?;
 
-        Ok(())
+                if !success {
+                    let error = format!(
+                        "Failed to recover the archiver for block number: {}",
+                        block_import_notification.block_number
+                    );
```

Comment:
Not sure we need to log an error here if we exit with it anyway. The whole node will crash and I believe we will log the error at higher level anyway.